lib/class-wp-duotone-gutenberg.php

```diff
@@ -0,0 +1,375 @@
+		if ( $has_duotone_attribute ) {
+
+			// Possible values for duotone attribute:
+			// 1. Array of colors - e.g. array('#000000', '#ffffff').
+			// 2. Variable for an existing Duotone preset - e.g. 'var:preset|duotone|green-blue' or 'var(--wp--preset--duotone--green-blue)''
+			// 3. A CSS string - e.g. 'unset' to remove globally applied duotone.
+
+			$duotone_attr = $block['attrs']['style']['color']['duotone'];
+			$is_preset    = is_string( $duotone_attr ) && WP_Duotone_Gutenberg::is_preset( $duotone_attr );
+			$is_css       = is_string( $duotone_attr ) && ! $is_preset;
+			$is_custom    = is_array( $duotone_attr );
+
+			if ( $is_preset ) {
+
+				// Extract the slug from the preset variable string.
+				$slug = WP_Duotone_Gutenberg::gutenberg_get_slug_from_attr( $duotone_attr );
+
+				// Utilize existing preset CSS custom property.
+				$declaration_value = WP_Duotone_Gutenberg::get_css_var( $slug );
+
+				WP_Duotone_Gutenberg::$output[ $slug ] = WP_Duotone_Gutenberg::$global_styles_presets[ $slug ];
+
+			} elseif ( $is_css ) {
+				// Build a unique slug for the filter based on the CSS value.
+				$slug = wp_unique_id( sanitize_key( $duotone_attr . '-' ) );
+
+				// Pass through the CSS value.
+				$declaration_value = $duotone_attr;
+			} elseif ( $is_custom ) {
+				// Build a unique slug for the filter based on the array of colors.
+				$slug = wp_unique_id( sanitize_key( implode( '-', $duotone_attr ) . '-' ) );
+
+				$filter_data = array(
+					'slug'   => $slug,
+					'colors' => $duotone_attr,
+				);
+				// Build a customized CSS filter property for unique slug.
+				$declaration_value = gutenberg_get_duotone_filter_property( $filter_data );
+
+				WP_Duotone_Gutenberg::$output[ $slug ] = $filter_data;
+			}
+		} elseif ( $has_global_styles_duotone ) {
+			$slug = WP_Duotone_Gutenberg::$global_styles_block_names[ $block['blockName'] ];
+
+			// Utilize existing preset CSS custom property.
+			$declaration_value = WP_Duotone_Gutenberg::get_css_var( $slug );
+
+			WP_Duotone_Gutenberg::$output[ $slug ] = WP_Duotone_Gutenberg::$global_styles_presets[ $slug ];
+		}
+
+		// - Applied as a class attribute to the block wrapper.
+		// - Used as a selector to apply the filter to the block.
+		$filter_id = gutenberg_get_duotone_filter_id( array( 'slug' => $slug ) );
+
+		// Build the CSS selectors to which the filter will be applied.
+		$selector = WP_Theme_JSON_Gutenberg::scope_selector( '.' . $filter_id, $duotone_support );
+
+		// We only want to add the selector if we have it in the output already, essentially skipping 'unset'.
+		if ( array_key_exists( $slug, WP_Duotone_Gutenberg::$output ) ) {
+			WP_Duotone_Gutenberg::$output[ $slug ]['selector'] = $selector;
+		}
+
+		// Calling gutenberg_style_engine_get_stylesheet_from_css_rules ensures that
+		// the styles are rendered in an inline for block supports because we're
+		// using the `context` option to instruct it so.
```

Comment:
> Do you think this comment is still relevant?

The way `gutenberg_style_engine_get_stylesheet_from_css_rules` is used hasn't been changed by this PR. It's still handling all of the class generation for duotone. So I assume it's still relevant.

> I'm not sure why we need to ensure it's in this inline block support style section.

I'm not sure why either. @andrewserong gave it the check when it was added, so he may be able to explain?

> Or is it that it only needs to be output after the global styles?

It's not really a matter of before or after in this particular case. It might help if I try to explain how this part of the code is related to global styles.

Most global styles have classes that can be applied to the block wrapper because the CSS properties they set are inherited properties. Global styles generates these based on the `'classes'` key in `PRESETS_METADATA` in [class-wp-theme-json-gutenberg.php](https://github.com/WordPress/gutenberg/blob/55e16d3c8a0e3b2ee8fd7f87a12cae8a2b1886d0/lib/class-wp-theme-json-gutenberg.php#L127). On every page, all of these class names are generated whether or not they are used. The mental model I have is that it works like all of the old-school CSS frameworks (Blueprint, Logic, YUI, etc.) that are just a single sheet of CSS that includes everything, but for entire blocks instead of just individual HTML elements.

Duotone is different because `filter` is a non-inherited CSS property which makes the CSS that needs to be generated dependent on the structure of each specific block. In hindsight I could have generated all the combinations of presets and registered block selectors to fit that paradigm. But that never even crossed my mind and it would have been even more wasteful than what what we're trying to fix now.

So what we ended up with was global styles classes for duotone being generated on block render. And that's why the function of this particular bit of code is still fulfilling the same role as it was prior to this PR.

Reply:
Thanks for the ping!

For a bit of context surrounding the `context` prop 😄 — prior to the style engine being introduced, block supports would output a style tag for each instance of a block being rendered, which resulted in numerous redundant style tags being output to the site frontend.

The style engine's `context` prop allows styles to be grouped together into a single stylesheet. In this case, I think it's fine to leave it with the `block-supports` context, as the style output for block supports is already handled, and unless there's a need to have a separate stylesheet for Duotone, I think it makes sense to combine into the one stylesheet. Separately, there are efforts being explored to roll out the style engine for global styles (e.g. in https://github.com/WordPress/gutenberg/pull/48955), but it should be pretty simple to change the `context` value here in a follow-up further down the track if we need to.

> Typo: Could we improve the "rendered in an inline" wording?

Perhaps it could be worded more like:

```
Pass styles to the block-supports stylesheet via the style engine. This ensures that Duotone styles are included in a single stylesheet, avoiding multiple style tags or multiple stylesheets being output to the site frontend.
```